Add retry logic for `upload_dir` in post-processing workflow

spencerkclark opened the pull request:
This PR adds retry logic via the `backoff` package to retry uploads via `gsutil rsync` if a `ResumableUploadException` occurs.  It does so by checking the output of the `gsutil` command if it fails and checking to see if the phrase `ResumableUploadException` appears.  

Requirement changes:
- Added `backoff` package as a dependency for the `post_process_run` image.
- [x] Ran `make lock_deps/lock_pip` following these [instructions](https://vulcanclimatemodeling.com/docs/fv3net/dependency_management.html#dependency-management)
- [x] Add PR review with license info for any additions to `constraints.txt`
  ([example](https://github.com/VulcanClimateModeling/fv3net/pull/1218#pullrequestreview-663644359))

- [x] Tests added

Resolves #1239

You are encouraged to check the PR against the [Code Review Checklist](https://paper.dropbox.com/doc/Code-Review-Checklist--A4lKrs~xg7w5Gsb39N6JLNQoAg-IlsYffZgTwyKEylty7NhY).

spencerkclark commented on constraints.txt:
```diff
@@ -23,6 +22,7 @@ attrs==19.3.0
 avro-python3==1.9.2.1
 babel==2.8.0
 backcall==0.2.0
+backoff==1.10.0
```
`backoff` uses an MIT license: https://github.com/litl/backoff/blob/master/LICENSE.

mcgibbon commented on workflows/post_process_run/fv3post/gsutil.py:
```diff
@@ -13,8 +26,19 @@ def authenticate():
         )
 
 
+@backoff.on_exception(backoff.expo, GSUtilResumableUploadException, max_tries=3)
```
Is there a reason for 3? How long do 3 retries take?

spencerkclark replied:
Given how infrequent the errors are, and that `rsync` should only retry the failed copies, I expect most of the time we will not need to retry more than twice.  Do you think I should increase it beyond 3?

mcgibbon commented on workflows/post_process_run/tests/test_gsutil.py:
```diff
@@ -0,0 +1,34 @@
+        ([UPLOAD_EXCEPTION_BYTES, None], None, 2),
+    ],
+)
+def test_upload_dir(monkeypatch, side_effect, exception, call_count):
```
The tests will currently pass if there's a bug in the code where the code checks that the message is equal to `"ResumableUploadException"`, instead of that it contains the exception. Can you add one case where the error message is a longer string containing the exception somewhere in the middle?

spencerkclark replied:
I modified the existing test cases such that they test this.

mcgibbon reviewed (approved):
LGTM! Should be reviewed by @nbren12 for licensing/dependency management. I'm surprised backoff wasn't already present, since I thought it was a dependency of fv3config or fv3gfs-wrapper, but it looks like not.

spencerkclark commented on constraints.txt:
```diff
@@ -44,6 +44,7 @@ colorcet==2.0.6
 commonmark==0.9.1
 conda-lock==0.4.1
 crcmod==1.7
+cryptography==3.4.7
```
`cryptography` uses either an Apache or BSD license: https://github.com/pyca/cryptography/blob/main/LICENSE.

spencerkclark commented on constraints.txt:
```diff
@@ -108,6 +109,7 @@ ipython-genutils==0.2.0
 ipython==7.16.1
 ipywidgets==7.6.3
 jedi==0.18.0
+jeepney==0.6.0
```
`jeepney` uses an MIT license: https://gitlab.com/takluyver/jeepney/-/blob/master/LICENSE.

mcgibbon commented:
Actually, it looks like if you do the review for dependencies of dependencies this is ready to merge, based on the PR checklist.

spencerkclark commented on constraints.txt:
```diff
@@ -209,6 +211,7 @@ retrying==1.3.3
 rsa==4.6
 scikit-learn==0.22.1
 scipy==1.5.4
+secretstorage==3.3.1
```
`secretstorage` uses a BSD license: https://github.com/mitya57/secretstorage/blob/master/LICENSE, https://github.com/mitya57/secretstorage/blob/aab3222745883fb854aa1f2a7a982a5259436259/setup.cfg#L13.

spencerkclark commented on constraints.txt:
```diff
@@ -248,6 +251,7 @@ tzlocal==2.1
 urllib3==1.25.10
 validators==0.18.2
 virtualenv==20.0.33
+watchdog==2.1.2
```
`watchdog` uses and Apache license: https://github.com/gorakhargosh/watchdog/blob/master/LICENSE.

spencerkclark commented:
> I'm surprised backoff wasn't already present, since I thought it was a dependency of fv3config or fv3gfs-wrapper, but it looks like not.

Yeah it used to be a dependency of fv3config, but [we removed it](https://github.com/VulcanClimateModeling/fv3config/pull/139) when we fixed the retry issues upstream in gcsfs.
